ls: move from getopts to clap and structure configuration

tertsdiepraam opened the pull request:
In the same vein as https://github.com/uutils/coreutils/issues/1795, https://github.com/uutils/coreutils/issues/1794 and https://github.com/uutils/coreutils/issues/1793, `ls` needs to use `clap` instead of `getopts`. That is what I originally set out to do, what I ended up doing was a much bigger refactor:
- Move from `getopts` to `clap` for argument handling.
- Create a `Config` struct which contains configuration options created from the command line arguments.
- Create enums  for:
  - Sorting: `None`, `Name`, `Size`, `Time` & `CTime`
  - Which files are displayed: `Normal`, `All` & `AlmostAll`
  - Size formats: `Bytes` (default) and `Binary` (specified by `--human-readable`,  this could have been a `bool`, but making this an enum makes adding `--si` easier later)
- Use the `Config` struct instead of the direct arguments in the rest of the program.
- Simplified the implementation of the `pad_left` function.

This moves a lot of awkward code for the configuration into a dedicated function and the use of enums makes adding new flags and options easier (especially `--si` and other sorting options such as `-v` & `-X`, which are sorting by version and extension, respectively).

sylvestre commented on src/uu/ls/src/ls.rs:
```diff
@@ -65,107 +69,291 @@ lazy_static! {
+        let sort = if options.is_present(options::SORT_TIME) {
+            Sort::Time
+        } else if options.is_present(options::SORT_CTIME) {
+            Sort::CTime
```
Could you please add a test for this case?

sylvestre commented on src/uu/ls/src/ls.rs:
```diff
@@ -65,107 +69,291 @@ lazy_static! {
+        };
+
+        let files = if options.is_present(options::ALL) {
+            Files::All
```
As the options seem to be exclusive, maybe following this example?
https://github.com/uutils/coreutils/blob/517b2a6a2244637a5f5331565592873a734085c7/src/uu/chown/src/chown.rs#L30

sylvestre commented on src/uu/ls/src/ls.rs:
```diff
@@ -65,107 +69,291 @@ lazy_static! {
+        let color = if options.is_present(options::COLOR) {
+            match options.value_of(options::COLOR) {
+                None => atty::is(atty::Stream::Stdout),
+                Some(val) => match val {
```
ditto for the test

sylvestre reviewed (changes requested):
Great work. Bravo!

Do you manage these cases?

$ ls --color 
$ ls --color=always file.txt
$ ls --color file.txt

I remember struggling with this pattern and blocked because of this:
https://github.com/clap-rs/clap/

tertsdiepraam commented:
Thank you for the review! I thought I implemented that behaviour correctly, but it seems I was wrong. It seems that `Arg::empty_values(true)` and `Arg::requires_equals(true)` do not play well together in `clap`. Switching to `Arg::min_values(0)` fixed it, along with another mistake I made. So it works correctly now. I'll update the PR with these changes and more tests tomorrow.

sylvestre commented:
I guess you saw that you have some warnings 

By the way, there is this crate:
https://crates.io/crates/lscolors
(but this isn't a request for this review to be clear :)
If you are interested, it should be done in a separate PR to avoid mixing too many things!

tertsdiepraam commented:
To recap what I've done since yesterday:
- Made the arguments overridable, i.e. `ls -l1` will give the single column format.
- Added `--sort`, `--format` and `--time`. They are only partially implemented, but work will the functionality that was already present via dedicated flags, i.e. `ls -l` will give the same result as `ls --format=long`.
- Fixed the `--color` flag such that `ls`, `ls --color=no`, etc. will not give colored output, but `ls --color`, `ls --color=yes` will. This also plays nicely with the format `ls --color file.txt`.
- `-c` and `-u` are now implemented not as sort options, but as time options and will show up in the output when `-l` is passed, just like in the GNU version.

One quirk remains, which is that the access time does not seem to be set when reading from a file on Windows. I don't have a Windows machine so I can't really check this apart from the output of the tests.

Finally the CI for Mac fails, but it seems unrelated to this PR.

The `lscolors` crate looks nice! I might look into that in a next PR.

sylvestre commented:
Please also add the --color various options ( https://github.com/uutils/coreutils/pull/1820#pullrequestreview-611720877 ) as test
(the three cases :)
thanks

next time, please don't mix changes in the same PR

tertsdiepraam commented:
Yes, that makes sense. I've expanded the tests for `--color`, which should cover all cases now and actually checks for colors.

I'll also keep my next PRs more contained!

sylvestre commented:
kudos!
